### PR TITLE
New loss functions and metrics

### PR DESCRIPTION
**Pull request by cjmcgill**
This PR changes the arguments handling of loss functions and metrics to allow for easier addition of new loss functions in the future. Separate files for loss functions and metrics are created in the chemprop/train directory to hold the respective custom functions. The structure of get_loss_function to more easily handle new loss functions.

Two new loss functions are added. The first is mcc (a soft version of matthews correlation coefficient) for classification and multiclass jobs. The second loss function is bounded_mse for regression jobs where some targets are simple inequalities (e.g., "<5" or similar). Supporting changes in loading data and passing inequality features to the loss and evaluation functions have been added.

Correspondingly, similar metrics have also been added. MCC and F1 metrics have been added for classification and multiclass. Bounded versions of mse, mae, and rmse have been added for regression metrics.

Some sections of the readme file have been rearranged to increase accessibility.

### CODE REVIEW

File: chemprop/train/loss_functions.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+def get_loss_func(args: TrainArgs) -> Callable:
+    """
+    Gets the loss function corresponding to a given dataset type.
+
+    :param args: Arguments containing the dataset type ("classification", "regression", or "multiclass").
+    :return: A PyTorch loss function.
+    """
+
+    # Nested dictionary of the form {dataset_type: {loss_function: loss_function callable}}
+    # Note f1 loss function has been disabled because mcc is similar but less sensitive to unbalanced datasets
```
Comment (kevingreenman) [inline chemprop/train/loss_functions.py:18]: If choosing to disable f1 here, should the functions below also be removed, and should it be removed from other files (e.g. args.py)? Or would it be better to allow it to be used with a warning/recommendation to use mcc instead?

Comment (cjmcgill) [reply chemprop/train/loss_functions.py:18]: It is still valid as a metric so I'm leaving it as a metric in args. I don't believe it's available in args as a loss function though.

I'm on the fence about removing it altogether from the loss function list. It was not insubstantial to figure out how to do it in the first place, before I had second thoughts about including it as a loss function. My thinking was that if anyone wanted to implement it later, this would allow them to easily. Now thinking about that weighted against having isolated code that wouldn't get updated with other functions, I think you're right. I'll go back and remove it.

Comment (cjmcgill) [reply chemprop/train/loss_functions.py]: Change made

**Review by hesther — COMMENTED** [review]
I ran a bunch of tests. Apart from the one error for multi class mcc loss, this seems to run fine. I manually calculated a few binary and multi class mcc scores and compared them to the function outputs, which reproduced the correct results. I also tested the bounded loss, and it seemed to work fine.

I like the restructuring to metrics.py and loss_functions.py, which makes the code easier to understand.

I am not sure what causes the lgtm alert. It seems to not like how the CSV is read in, since it comes from a user-provided path which potentially could attackers lead to read in malicious files - this is probably ok to ignore from my point of view.

File: chemprop/train/loss_functions.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    :return: A tensor containing loss values for the batch with shape (batch_size,spectrum_length).
+    """
+    # Move new tensors to torch device
+    torch_device = model_spectra.device
+
+    # Normalize the model spectra before comparison
+    zero_sub = torch.zeros_like(model_spectra, device=torch_device)
+    one_sub = torch.ones_like(model_spectra, device=torch_device)
+    if threshold is not None:
+        threshold_sub = torch.full(model_spectra.shape,threshold, device=torch_device)
```
[the remark this answers is not in the document]
Comment (cjmcgill) [reply chemprop/train/loss_functions.py:124]: Good change

File: chemprop/train/loss_functions.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        raise ValueError(f'Dataset type "{args.dataset_type}" not supported.')
+
+    # Return loss function if it is represented in the supported_loss_functions dictionary
+    loss_function = supported_loss_functions.get(args.dataset_type, dict()).get(args.loss_function, None)
+
+    if loss_function is not None:
+        return loss_function
+
+    else:
+        raise ValueError(f'Loss function "{args.loss_function}" not supported with dataset type {args.dataset_type}. Available options for that dataset type are {supported_loss_functions[args.dataset_type].keys()}.')
```
[the remark this answers is not in the document]
Comment (cjmcgill) [reply chemprop/train/loss_functions.py:48]: Good suggestion

**Review by kevingreenman — APPROVED** [review]
This looks good to me other than the minor suggestion about f1 that you fixed. I like the new structure with the separate files for loss functions and metrics.

File: chemprop/train/train.py [hunk 4]
```diff
@@ -68,13 +74,23 @@ def train(model: MoleculeModel,
...
+            lt_target_batch = lt_target_batch.to(torch_device)
+            gt_target_batch = gt_target_batch.to(torch_device)
 
         # Calculate losses
-        if args.dataset_type == 'multiclass':
+        if args.loss_function == 'mcc' and args.dataset_type == 'classification':
+            loss = loss_func(preds, targets, data_weights, mask) *target_weights.squeeze(0)
+        elif args.loss_function == 'mcc': # multiclass dataset type
+            targets = targets.long()
+            loss = torch.cat([loss_func(preds[:, target_index, :], targets[:, target_index], data_weights[:, target_index], mask[:, target_index]) for target_index in range(preds.size(1))], dim=1) * target_weights.squeeze(0)
```
[the remark this answers is not in the document]
Comment (cjmcgill) [reply chemprop/train/train.py:86]: This is an important catch. Multiclass is not currently included in the integration testing. And the manual testing I did was just on the function and not on this aggregation step. Thanks for the catch and the fix!

[the remark this answers is not in the document]
Comment (cjmcgill) [reply chemprop/train/train.py]: I tried it with a test file of my own that had two tasks for the multiclass job. And it revealed another problem. To find the source of the dimensionality problem, I had to break up the multiclass loss into multiple lines. And since I've always kind of thought the multiclass loss line was compact but unreadable, I unfolded it for the regular multiclass case too.

**Review by hesther — APPROVED** [review]
Looks good to me!
